src/devices/Media/SoundDevice/Devices/UnixSoundDevice.cs

```diff
@@ -177,114 +207,156 @@ public override void Record(uint recordTimeSeconds, Stream outputStream)
                 SubChunk2 = subChunk2
             };
 
-            WriteWavHeader(outputStream, header);
+            return header;
+        }
 
-            try
+        public override void StartRecording(string outputFilePath)
+        {
+            Record(outputFilePath);
+        }
+
+        public override void StopRecording()
+        {
+            _record = false;
+            while (_recordingThread.IsAlive)
             {
-                OpenRecordingPcm();
-                PcmInitialize(_recordingPcm, header, ref @params, ref dir);
-                ReadStream(outputStream, header, ref @params, ref dir);
-                CloseRecordingPcm();
+                Thread.Sleep(1);
             }
-            finally
+        }
+
+        private void Record(string outputFilePath)
+        {
+            _recordingThread = new Thread(() =>
             {
-                Dispose();
-            }
+                IntPtr @params = new IntPtr();
+                int dir = 0;
+                _recordingTotalTimeSeconds = 0;
+                var header = CreateHeader(1);
+                using (var recordingFile = File.Open(outputFilePath, FileMode.Create))
+                {
+                    try
+                    {
+                        OpenRecordingPcm();
+                        PcmInitialize(_recordingPcm, header, ref @params, ref dir);
+                        _record = true;
+                        while (_record)
+                        {
+                            _recordingTotalTimeSeconds++;
+                            ReadStream(recordingFile, header, ref @params, ref dir);
+                        }
+
+                        CloseRecordingPcm();
+                    }
+                    finally
+                    {
+                        Dispose();
+                    }
+
+                    // Write header with total time
+                    header = CreateHeader(_recordingTotalTimeSeconds);
+                    recordingFile.Position = 0;
+                    WriteWavHeader(recordingFile, header);
+                    recordingFile.Flush();
+                    recordingFile.Close();
+                    recordingFile.Dispose();
+                }
+            });
+            _recordingThread.Start();
         }
 
-        private void WriteWavHeader(Stream wavStream, WavHeader header)
+        public override void WriteWavHeader(Stream wavStream, WavHeader header)
         {
             Span<byte> writeBuffer2 = stackalloc byte[2];
             Span<byte> writeBuffer4 = stackalloc byte[4];
 
-            Encoding.ASCII.GetBytes(header.Chunk.ChunkId, writeBuffer4);
-            wavStream.Write(writeBuffer4);
+            writeBuffer4 = Encoding.ASCII.GetBytes(header.Chunk.ChunkId);
+            wavStream.Write(writeBuffer4.ToArray(), 0, writeBuffer4.Length);
```

Comment:
this is allocating quite a bit... You probably don't benefit much fron using Span here. Actually might be harmful since you need to convert Array -> Span -> Array. In .net5 you got Span overloads for Write